server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -168,6 +170,7 @@ public String toString() {
             if (event.localNodeMaster() == false) {
                 onNoLongerMaster();
             }
+            desiredBalanceMetrics.setNodeIsMaster(event.localNodeMaster());
```

Comment:
Nit: I still think checking `event.localNodeMaster() != event.previousState().nodes().isLocalNodeElectedMaster()` before calling this method is helpful since otherwise most of the time it ends up setting the same value.